File: C++/Medium/WrongAnswer/minimum-cost-for-tickets_4.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        
        int n = days.size();
        int dp[n];
        dp[0] = min({costs[0], costs[1], costs[2]});
        for (int i = 1; i < n; i ++) {
            dp[i] = dp[i - 1] + costs[0];
            for (int j = i - 1; j >= 0; j --) {
                if (days[i] - days[j] < 7) {
                    dp[i] = min(dp[i], dp[j] + costs[1]);
                    break;
                }
            }
            for (int j = i - 1; j >= 0; j --) {
                if (days[i] - days[j] < 30) {
                    dp[i] = min(dp[i], dp[j] + costs[2]);
                    break;
                }
            }
        }
        return dp[n - 1];
    }
};

File: Java/Medium/Accepted/minimum-cost-for-tickets_4.java
class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        
        int[] dp = new int[days.length];
        Arrays.fill(dp, -1);
        return mincostTickets(days, costs, 0, dp);
    }
    public int mincostTickets(int[] days, int[] costs, int i, int[] dp) {
        if (i >= days.length) {
            return 0;
        }
        if (dp[i] != -1) {
            return dp[i];
        }
        int min = Integer.MAX_VALUE;
        int j = i;
        min = Math.min(min, costs[0] + mincostTickets(days, costs, i + 1, dp));
        while (j < days.length && days[j] < days[i] + 7) {
            j++;
        }
        min = Math.min(min, costs[1] + mincostTickets(days, costs, j, dp));
        while (j < days.length && days[j] < days[i] + 30) {
            j++;
        }
        min = Math.min(min, costs[2] + mincostTickets(days, costs, j, dp));
        dp[i] = min;
        return min;
    }
}